Add new totp sample

rajdeepnanua-okta opened the pull request:
Still have to add unit tests

But, I decided to put up a draft PR to show the general approach for the sample. This app was also tested by scanning the same QR code in both Google authenticator, and totp app. Both generated the same results for the default period, digits, and algorithm parameters. This app also allows customizing period, digits, and algorithm. I couldn't find a good test with those parameters as non-default values.

rajdeepnanua-okta commented on build.gradle:
```diff
@@ -14,6 +17,11 @@ buildscript {
 }
 
+plugins {
+    id 'com.google.dagger.hilt.android' version '2.44' apply false
+    id 'org.jetbrains.kotlin.android' version '1.7.10' apply false
```
I couldn't find a way to use the versions in libs.versions.toml here

rajdeepnanua-okta commented on totp/build.gradle:
```diff
@@ -1,35 +1,81 @@
+    implementation(libs.bundles.android)
+    implementation(libs.bundles.compose.ui)
+    implementation(libs.hilt.android)
+    implementation 'com.google.android.gms:play-services-mlkit-barcode-scanning:18.1.0'
+    implementation 'androidx.security:security-crypto-ktx:1.1.0-alpha03'
+    def camerax_version = "1.2.0-beta02"
+    // The following line is optional, as the core library is included indirectly by camera-camera2
+    implementation "androidx.camera:camera-core:${camerax_version}"
+    implementation "androidx.camera:camera-camera2:${camerax_version}"
+    // If you want to additionally use the CameraX Lifecycle library
+    implementation "androidx.camera:camera-lifecycle:${camerax_version}"
+    // If you want to additionally use the CameraX VideoCapture library
+    implementation "androidx.camera:camera-video:${camerax_version}"
+    // If you want to additionally use the CameraX View class
+    implementation "androidx.camera:camera-view:${camerax_version}"
+    // If you want to additionally use the CameraX Extensions library
+    implementation "androidx.camera:camera-extensions:${camerax_version}"
+    implementation "androidx.camera:camera-mlkit-vision:${camerax_version}"
+    implementation "com.google.accompanist:accompanist-permissions:0.25.1"
+    implementation 'dev.turingcomplete:kotlin-onetimepassword:2.4.0'
+    implementation("com.google.guava:guava:31.1-jre")
+    implementation "androidx.navigation:navigation-compose:2.5.2"
+    implementation "androidx.hilt:hilt-navigation-compose:1.0.0"
```
These will be moved to libs.version.toml for the non-draft PR

JayNewstrom commented on totp/src/main/java/com/okta/totp/otp_display/PasswordGenerator.kt:
```diff
@@ -0,0 +1,28 @@
+    )
+
+    fun generate(): String {
+        GoogleAuthenticator
```
This looks out of place.

rajdeepnanua-okta replied:
Whoops, yep that was by accident

Jenmaddox commented:
Why do I keep getting these emails?

***@***.***


On Oct 20, 2022, at 11:15 AM, Jay Newstrom ***@***.***> wrote:

﻿

@JayNewstrom commented on this pull request.

________________________________

In totp/src/main/java/com/okta/totp/ui/theme/Theme.kt<https://github.com/okta/samples-android/pull/91#discussion_r1000814421>:

> +    primary = Purple500,
+    primaryVariant = Purple700,
+    secondary = Teal200
+
+    /* Other default colors to override
+    background = Color.White,
+    surface = Color.White,
+    onPrimary = Color.White,
+    onSecondary = Color.Black,
+    onBackground = Color.Black,
+    onSurface = Color.Black,
+    */
+)
+
***@***.***
+fun SamplesandroidTheme(


Can you rename this to SamplesAndroidTheme

________________________________

In totp/src/main/java/com/okta/totp/ui/theme/Theme.kt<https://github.com/okta/samples-android/pull/91#discussion_r1000814759>:

> +import androidx.compose.material.darkColors
+import androidx.compose.material.lightColors
+import androidx.compose.runtime.Composable
+
+private val DarkColorPalette = darkColors(
+    primary = Purple200,
+    primaryVariant = Purple700,
+    secondary = Teal200
+)
+
+private val LightColorPalette = lightColors(
+    primary = Purple500,
+    primaryVariant = Purple700,
+    secondary = Teal200
+
+    /* Other default colors to override


We should remove commented out code (there are a few other places like this as well).

________________________________

In totp/src/main/java/com/okta/totp/ui/theme/Color.kt<https://github.com/okta/samples-android/pull/91#discussion_r1000816539>:

> @@ -0,0 +1,8 @@
+package com.okta.totp.ui.theme


Need to add copyright headers.

________________________________

In totp/src/main/java/com/okta/totp/barcode_scan/BarcodeScanScreen.kt<https://github.com/okta/samples-android/pull/91#discussion_r1000818923>:

> +import androidx.core.content.ContextCompat
+import com.google.accompanist.permissions.ExperimentalPermissionsApi
+import com.google.accompanist.permissions.PermissionStatus
+import com.google.accompanist.permissions.rememberPermissionState
+import com.google.accompanist.permissions.shouldShowRationale
+import com.google.mlkit.vision.barcode.BarcodeScanner
+import com.google.mlkit.vision.barcode.BarcodeScanning
+
***@***.***
+fun BarcodeScanScreen(barcodeScanViewModel: BarcodeScanViewModel, onNavigateToOtpScreen: () -> Unit) {
+    Scaffold(
+        topBar = {
+            Row(modifier = Modifier.background(color = Color.LightGray)) {
+                Icon(
+                    imageVector = Icons.Default.ArrowBack,
+                    contentDescription = "Go Back",


We should add these to strings.xml (there's a lot of these, but I'll just tag this one).

________________________________

In totp/src/main/java/com/okta/totp/barcode_scan/BarcodeScanScreen.kt<https://github.com/okta/samples-android/pull/91#discussion_r1000821732>:

> +import com.google.accompanist.permissions.rememberPermissionState
+import com.google.accompanist.permissions.shouldShowRationale
+import com.google.mlkit.vision.barcode.BarcodeScanner
+import com.google.mlkit.vision.barcode.BarcodeScanning
+
***@***.***
+fun BarcodeScanScreen(barcodeScanViewModel: BarcodeScanViewModel, onNavigateToOtpScreen: () -> Unit) {
+    Scaffold(
+        topBar = {
+            Row(modifier = Modifier.background(color = Color.LightGray)) {
+                Icon(
+                    imageVector = Icons.Default.ArrowBack,
+                    contentDescription = "Go Back",
+                    modifier = Modifier
+                        .clickable {
+                            onNavigateToOtpScreen()


The syntax for this can be simplified to remove the extra function reference: .clickable(onClick = onNavigateToOtpScreen)

________________________________

In totp/src/main/java/com/okta/totp/barcode_scan/BarcodeScanScreen.kt<https://github.com/okta/samples-android/pull/91#discussion_r1000829115>:

> +                    "The camera is important for this app. Please grant the permission."
+                } else {
+                    "Camera permission required for this feature to be available. " +
+                            "Please grant the permission"
+                }
+                Text(textToShow)
+                Button(onClick = { cameraPermissionState.launchPermissionRequest() }) {
+                    Text("Request permission")
+                }
+            }
+        }
+    }
+}
+
***@***.***
+fun CameraPreview(


This code looks like it's quite prone to errors and/or leaking. Please run some manual tests here relating to app switching, rotation, and screen on/off.

________________________________

In totp/src/main/java/com/okta/totp/parsing/OtpUriParser.kt<https://github.com/okta/samples-android/pull/91#discussion_r1000847460>:

> +                name = label[1].trim('/')
+            }
+            else -> return OtpUriParsingResults.Error(
+                "Totp URI has invalid label"
+            )
+        }
+
+        val paramNames = uri.queryParameterNames
+        if (!paramNames.contains(SECRET)) {
+            return OtpUriParsingResults.Error(
+                "OTP URI is missing required parameter: secret"
+            )
+        }
+
+        if (paramNames.contains(ISSUER)) {
+            val issuerParam = uri.getQueryParameter(ISSUER)!!


This isn't safe. The query param could have no value, even if it exists as a key. foo=&bar=with_a_value. It might be better to do uri.getQueryParameter(ISSUER)?.let { to prevent an error like this.

—
Reply to this email directly, view it on GitHub<https://github.com/okta/samples-android/pull/91#pullrequestreview-1149524244>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ARQAW2QJU3ZKS653SZEPJGDWEFV2DANCNFSM6AAAAAARJISQTI>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

JayNewstrom commented:
@Jenmaddox you might be watching the repo, or subscribed to the PR. You can unwatch the repo at the top of the page.

rajdeepnanua-okta commented on totp/src/main/java/com/okta/totp/barcode_scan/BarcodeScanScreen.kt:
```diff
@@ -0,0 +1,205 @@
+            Row(modifier = Modifier.background(color = Color.LightGray)) {
+                Icon(
+                    imageVector = Icons.Default.ArrowBack,
+                    contentDescription = "Go Back",
```
Agreed. I will do that

rajdeepnanua-okta commented on totp/src/main/java/com/okta/totp/parsing/OtpUriParser.kt:
```diff
@@ -0,0 +1,112 @@
+        }
+
+        if (paramNames.contains(ISSUER)) {
+            val issuerParam = uri.getQueryParameter(ISSUER)!!
```
Good call. I'll fix that for secret as well

rajdeepnanua-okta commented:
I've addressed the comments and added most of the unit tests. Still finishing up unit tests for OtpDisplayViewModel, and will convert the PR to non-draft once that's done

rajdeepnanua-okta commented on totp/src/main/java/com/okta/totp/otp_display/OtpDisplayScreen.kt:
```diff
@@ -0,0 +1,195 @@
+        return super.equals(other)
+    }
+
+    override fun hashCode(): Int {
+        var result = otpCode.hashCode()
+        result = 31 * result + account.hashCode()
+        result = 31 * result + (issuer?.hashCode() ?: 0)
+        return result
+    }
```
This code was autogenerated by the IDE. I'm open to switching to using Objects.hash if that's preferable

rajdeepnanua-okta commented on totp/src/main/java/com/okta/totp/otp_display/OtpDisplayViewModel.kt:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    @VisibleForTesting
+    fun setMaxUpdateCycles(maxUpdates: Int) {
+        limitUpdates = true
+        this.maxUpdates = maxUpdates
+    }
```
Had to add this so that init didn't run forever. I'm using runTest for OtpDisplayViewModelTest, and the test coroutine there automatically skips delays. I couldn't find a way to disable the delay-skipping behavior, so I've added this for now until I find a better way

JayNewstrom commented on totp/src/main/java/com/okta/totp/otp_display/OtpDisplayScreen.kt:
```diff
@@ -0,0 +1,195 @@
+    )
+}
+
+data class OtpEntry(
```
Probably makes sense to remove the data modifier is we're doing custom equals/hashcode.

JayNewstrom reviewed (approved):
Overall looks good.
